core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PrioritizeUtilizationExecutionSchedule.java

```diff
@@ -0,0 +1,559 @@
+        });
+
+        Visitor visitor = new Visitor(fragments);
+        fragments.forEach(fragment -> visitor.processFragment(fragment.getId()));
+
+        // Make sure there are no strongly connected components as it would mean circular dependency between stages
+        List<Set<PlanFragmentId>> components = new StrongConnectivityInspector<>(fragmentDependency).stronglyConnectedSets();
+        verify(components.size() == fragmentDependency.vertexSet().size(), "circular dependency between stages");
+
+        return visitor.getNonLazyFragments();
+    }
+
+    private class Visitor
+            extends PlanVisitor<FragmentSubGraph, PlanFragmentId>
+    {
+        private final Map<PlanFragmentId, PlanFragment> fragments;
+        private final ImmutableSet.Builder<PlanFragmentId> nonLazyFragments = ImmutableSet.builder();
+        private final Map<PlanFragmentId, FragmentSubGraph> fragmentSubGraphs = new HashMap<>();
+
+        public Visitor(Collection<PlanFragment> fragments)
+        {
+            this.fragments = requireNonNull(fragments, "fragments is null").stream()
+                    .collect(toImmutableMap(PlanFragment::getId, identity()));
+        }
+
+        public Set<PlanFragmentId> getNonLazyFragments()
+        {
+            return nonLazyFragments.build();
+        }
+
+        public FragmentSubGraph processFragment(PlanFragmentId planFragmentId)
+        {
+            if (fragmentSubGraphs.containsKey(planFragmentId)) {
+                return fragmentSubGraphs.get(planFragmentId);
+            }
+
+            FragmentSubGraph subGraph = processFragment(fragments.get(planFragmentId));
+            fragmentSubGraphs.put(planFragmentId, subGraph);
+            return subGraph;
+        }
+
+        private FragmentSubGraph processFragment(PlanFragment fragment)
+        {
+            FragmentSubGraph subGraph = fragment.getRoot().accept(this, fragment.getId());
+            // append current fragment to set of upstream fragments as it is no longer being visited
+            Set<PlanFragmentId> upstreamFragments = ImmutableSet.<PlanFragmentId>builder()
+                    .addAll(subGraph.getUpstreamFragments())
+                    .add(fragment.getId())
+                    .build();
+            Set<PlanFragmentId> lazyUpstreamFragments;
+            if (subGraph.isCurrentFragmentLazy()) {
+                // append current fragment as a lazy fragment as it is no longer being visited
+                lazyUpstreamFragments = ImmutableSet.<PlanFragmentId>builder()
+                        .addAll(subGraph.getLazyUpstreamFragments())
+                        .add(fragment.getId())
+                        .build();
+            }
+            else {
+                lazyUpstreamFragments = subGraph.getLazyUpstreamFragments();
+                nonLazyFragments.add(fragment.getId());
+            }
+            return new FragmentSubGraph(upstreamFragments, lazyUpstreamFragments, false);
```

Comment:
why always `false`? Shouldn't it be `true/false` depending on the if branch above?

Reply:
It's no longer relevant as we have finished visiting current fragment. `FragmentSubGraph#currentFragmentLazy` is a support field when traversing current fragment. We only learn if current fragment is lazy or not when we finish visiting it.

Added comment